File: 0040-combination-sum-ii/0040-combination-sum-ii.cpp
class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        vector<vector<int>> ans;
        vector<int> temp;
        sort(candidates.begin(),candidates.end());
        helper(0,temp,candidates,ans,target);
        
        return ans;
    }
    
    void helper(int idx, vector<int>& temp,vector<int>& candidates,vector<vector<int>>& ans,int target){
        
        if(target == 0){
            ans.push_back(temp);
            return;
        }
        
        for(int i=idx;i<candidates.size();i++){
            if(i>idx && candidates[i] == candidates[i-1]) continue;
            if(candidates[i] > target) break;
            
            temp.push_back(candidates[i]);
            helper(i+1,temp,candidates,ans,target-candidates[i]);
            temp.pop_back();
        }
    }
};

File: 0040-combination-sum-ii/0040-combination-sum-ii.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        ans = []
        candidates.sort()

        def dfs(idx,curSum,arr):
            if curSum == target:
                ans.append(arr.copy())
                return
            
            if idx == len(candidates) or curSum > target:
                return

            arr.append(candidates[idx])
            dfs(idx+1,curSum + candidates[idx],arr)
            arr.pop()


            while idx+1 < len(candidates) and candidates[idx] == candidates[idx+1]:
                idx = idx + 1

            dfs(idx+1,curSum,arr)


        dfs(0,0,[])

        return ans